pkg/pool/pool.go

```diff
@@ -13,17 +13,35 @@
 
 package pool
 
-import "sync"
+import (
+	"errors"
+	"reflect"
+	"sync"
+)
 
-// BytesPool is a bucketed pool for variably sized byte slices.
-type BytesPool struct {
-	buckets []sync.Pool
-	sizes   []int
+type bucket struct {
+	slices []sync.Pool
+	// new is the function used to initialise an empty slice when none exist yet.
+	new func(int) interface{}
+	// Holds all available sizes for this bucket.
+	sizes []int
 }
 
-// NewBytesPool returns a new BytesPool with size buckets for minSize to maxSize
-// increasing by the given factor.
-func NewBytesPool(minSize, maxSize int, factor float64) *BytesPool {
+// Pool is a bucketed pool for variably sized slices of any type.
+type Pool struct {
+	buckets map[interface{}]bucket
+}
+
+// New initializes a new pool.
+func New() *Pool {
+	p := &Pool{
+		buckets: make(map[interface{}]bucket),
+	}
+	return p
+}
+
+// Add adds a new bucket with a given id to the Pool with size for minSize to maxSize increasing by the given factor.
+func (p *Pool) Add(id interface{}, minSize, maxSize int, factor float64, newFunc func(int) interface{}) {
```

Comment:
A pool with similar semantics for arbitrary slice types is certainly useful. But I'm not sure we also need to wrap this once again to have an object holding multiple of those pools.

The naming is also rather confusing as the former Pool is now called a bucket – but a bucket for a certain type/id (?) now rather than a certain slice size range.

I would make it a bit simpler again and really just move the existing bucket from `[]byte` to []interface{}`.

Reply:
classic over-engineering  :)

please check the updated version and let me know if you see anyhow to simplify it..